md-proto/WebAssembly.md

```diff
@@ -0,0 +1,2037 @@
+
+> Validation types are only used during validation, and not during execution.
+
+> `any` is effectively serving as a kind of bottom type here, and `void` is
+effectively serving as a kind of top type here. Note that in some type theory
+contexts, it's common to use these names for exactly opposite purposes.
+
+#### Validation Type Merge
+
+To merge two validation types:
+ - If either validation type is `any`, the result is the other validation type.
+ - Otherwise, if the validation types are the same, the result is that
+   validation type.
+ - Otherwise, validation fails.
+
+### Module Validation
+
+TODO: Describe module validation.
+
+### Function Validation
+
+For the duration of the validation of a function body, several data structures
+are created:
+ - A *control-flow stack*, with entries each containing
+    - A [validation type](#validation-type).
+    - A *limit* value.
+ - A *[type] stack*.
+ - A *current position*.
+
+During function validation, if either stack is empty when an element is to be
+popped from it, or an element is accessed by index less than zero or not less
+than the stack's length, validation fails. If a [validation type] is popped from
+the type stack such that the length of the type stack becomes less than the
+control-flow stack top's limit value, validation fails.
+
+#### Function Validation Initialization
+
+If the instruction sequence is empty, or if the last instruction in the sequence
+is not an `end`, validation fails.
+
+The current position starts at the first position. The type stack begins empty.
+The control-flow stack starts with one entry. This entry's [validation type] is
+the first return type of the function, if it has any, or `void` otherwise. Its
+limit value is zero.
+
+> The final `end` instruction may be implicit in some representations.
+
+#### Function Body Validation
+
+The instruction at the current position is remembered, and the current position
+is incremented to point to the position following it. Then the remembered
+instruction is validated as follows.
+
+If the instruction has a **Validation** clause, that clause describes its
+validation. Otherwise, [generic validation](#generic-instruction-validation) is
+performed.
+
+If the current position is now past the end of the sequence,
+[function return validation](#function-return-validation) is initiated and
+validation of the function is thereafter complete.
+
+Otherwise, [validation](#function-body-validation) is restarted with the new
+current position.
+
+#### Generic Instruction Validation
+
+For each operand [type] in the instruction's signature in reverse order, a
+[validation type] is popped from the type stack and checked to match it. Each
+return type of the instruction's signature is then pushed onto the type stack.
+
+If the instruction's signture has no return types:
+ - If the length of the type stack is greater than the control-flow stack top's
+   limit value, validation fails.
```

Comment:
I've now removed all language which started to anticipate multiple return values.

The `(nop) (nop)` problem is interesting. This is a consequence of the new explicit drop rule which I hadn't yet discovered. Real implementations will presumably need to solve this too. I added a TODO about it. It may be that a no-push-void approach is a good way to handle this, but no-push-void is still experimental in other ways, and I don't want it to be a distraction here.